staging/src/k8s.io/apiserver/pkg/endpoints/responsewriter/wrapper.go

```diff
@@ -0,0 +1,143 @@
+	Unwrap() http.ResponseWriter
+}
+
+// Wrap accepts a Decorator object and wraps with an "outer" http.ResponseWriter
+// object that implements http.CloseNotifier, http.Flusher and http.Hijacker accordingly
+func Wrap(decorator Decorator) http.ResponseWriter {
+	inner := decorator.Unwrap()
+
+	_, notifiable := inner.(http.CloseNotifier)
+	_, hijackable := inner.(http.Hijacker)
+	_, flushable := inner.(http.Flusher)
+
+	// from go net/http documentation:
+	// The default HTTP/1.x and HTTP/2 ResponseWriter implementations support Flusher
+	// Handlers should always test for this ability at runtime.
+	//
+	// The Hijacker interface is implemented by ResponseWriters that allow an HTTP handler
+	// to take over the connection.
+	// The default ResponseWriter for HTTP/1.x connections supports Hijacker, but HTTP/2 connections
+	// intentionally do not. ResponseWriter wrappers may also not support Hijacker.
+	// Handlers should always test for this ability at runtime
+	//
+	// The CloseNotifier interface is implemented by ResponseWriters which allow detecting
+	// when the underlying connection has gone away.
+	// Deprecated: the CloseNotifier interface predates Go's context package.
+	// New code should use Request.Context instead.
+	switch {
+	case notifiable && hijackable && flushable:
+		// for HTTP/1.x request the default implementation of ResponseWriter
+		// also implement CloseNotifier, Flusher and Hijacker
+		return &withCloseNotifyFlushAndHijack{
+			withCloseNotifyAndFlush: &withCloseNotifyAndFlush{
+				Decorator: decorator,
+			},
+		}
+	case notifiable && flushable:
+		// for HTTP/2 request, the default ResponseWriter object (http2responseWriter)
+		// implements Flusher and CloseNotifier.
+		return &withCloseNotifyAndFlush{
+			Decorator: decorator,
+		}
+	default:
+		// we should never be here for either http/1.x or http2 request
+		return decorator
+	}
+}
+
+// GetOriginal goes through the chain of wrapped http.ResponseWriter objects
+// and returns the original http.ResponseWriter object provided by net/http
+func GetOriginal(w http.ResponseWriter) http.ResponseWriter {
+	decorator, ok := w.(Decorator)
+	if !ok {
+		return w
+	}
+
+	inner := decorator.Unwrap()
+	if inner == w {
+		// infinite cycle here, we should never be here though.
+		panic("http.ResponseWriter decorator chain has a cycle")
+	}
+
+	return GetOriginal(inner)
+}
+
+//lint:ignore SA1019 backward compatibility
+var _ http.CloseNotifier = &withCloseNotifyAndFlush{}
+var _ http.Flusher = &withCloseNotifyAndFlush{}
+var _ http.ResponseWriter = &withCloseNotifyAndFlush{}
+var _ Decorator = &withCloseNotifyAndFlush{}
+
+type withCloseNotifyAndFlush struct {
+	Decorator
+}
+
+func (wr *withCloseNotifyAndFlush) CloseNotify() <-chan bool {
+	if notifier, ok := wr.Decorator.(http.CloseNotifier); ok {
+		return notifier.CloseNotify()
+	}
+
+	inner := wr.Decorator.Unwrap()
+	return inner.(http.CloseNotifier).CloseNotify()
```

Comment:
We could avoid the type cast here by giving `withCloseNotifyAndFlush` a field holding the inner RW typed by an interface that is the combination of `http.ResponseWriter`, `http.CloseNotifier`, and `http.Flusher`.

Reply:
done